components/locale_core/src/preferences/mod.rs

```diff
@@ -4,238 +4,626 @@
-                use $crate::preferences::PreferenceKey;
+                $name::from(&loc)
+            }
+        }
 
-                let lid = Some(loc.id);
+        impl From<&$crate::Locale> for $name {
+            fn from(loc: &$crate::Locale) -> Self {
+                use $crate::preferences::PreferenceKey;
 
                 $(
                     let mut $key = None;
                 )*
 
                 for (k, v) in loc.extensions.unicode.keywords.iter() {
                     $(
-                      if <$pref>::matches_ue_key(k) {
-                          if let Ok(r) = TryInto::<$pref>::try_into(v.clone()) {
-                              if !r.is_custom() {
-                                  $key = Some(r);
-                              }
-                          }
-                      }
+                        if let Ok(Some(r)) = <$pref>::try_from_key_value(k, v) {
+                            $key = Some(r);
+                            continue;
+                        }
                     )*
                 }
 
+                let sd = loc
+                    .extensions
+                    .unicode
+                    .keywords
+                    .get(&$crate::extensions::unicode::key!("sd"))
+                    .and_then(|v| v.as_single_subtag().copied());
+                let ue_region = loc
+                        .extensions
+                        .unicode
+                        .keywords
+                        .get(&$crate::extensions::unicode::key!("rg"))
+                        .and_then(|v| v.as_single_subtag()
+                            .and_then(|s| $crate::subtags::Region::try_from_str(s.as_str()).ok()));
                 Self {
-                    lid,
+                    language: loc.id.language,
+                    script: loc.id.script,
+                    region: loc.id.region,
+                    variant: loc.id.variants.iter().copied().next(),
+                    subdivision: sd,
+                    ue_region,
+
                     $(
                         $key,
                     )*
                 }
             }
         }
 
+        impl From<$crate::LanguageIdentifier> for $name {
+            fn from(lid: $crate::LanguageIdentifier) -> Self {
+                $name::from(&lid)
+            }
+        }
+
+        impl From<&$crate::LanguageIdentifier> for $name {
+            fn from(lid: &$crate::LanguageIdentifier) -> Self {
+                Self {
+                    language: lid.language,
+                    script: lid.script,
+                    region: lid.region,
+                    variant: lid.variants.iter().copied().next(),
+                    subdivision: None,
+                    ue_region: None,
+
+                    $(
+                        $key: None,
+                    )*
+                }
+            }
+        }
+
+        impl From<&icu_provider::DataLocale> for $name {
```

Comment:
DurationFormatter takes `DataLocale` and I need to get preferences out of it.

Reply:
DurationFormatter should take DurationFormatterPreferences, no?

Reply:
Once we migrate it, yeah. But for now, I'm narrowing down the scope of this patch and would prefer to avoid scope creep. I'm landing an impl and a migration of list formatter, and that requires me to massage the duration formatter.

Reply:
The macro shouldn't generate this for all preferences, the downstream crate can add a temporary implementation. There should also be a TODO because we should definitely not launch with this.